cli_tools/osconfig_agent/patch_window.go

```diff
@@ -15,26 +15,306 @@
+}
+
+// register registers a patch window as active, this will clobber any existing
+// window with the same name.
+func (w *patchWindow) register() {
+	fmt.Println("DEBUG: register", w.Name)
+	w.mx.Lock()
+	defer w.mx.Unlock()
+	fmt.Println("DEBUG: register got lock", w.Name)
+	defer func() { fmt.Println("DEBUG: register unlock", w.Name) }()
+
+	aw.add(w.Name, w)
+
+	// Create the Timer that will kick off the patch process.
+	// If we happen to be in the patch window now this will start imediately.
+	w.cancel = make(chan struct{})
+	w.t = time.AfterFunc(w.Start.Sub(time.Now()), func() { rc <- w })
+	fmt.Println(w.Name, "register done")
+}
+
+// deregister stops a patch windows timer and removes it from activeWindows
+// but does not cancel any current runs.
+func (w *patchWindow) deregister() {
+	fmt.Println(w.Name, "deregister")
+	w.t.Stop()
+	aw.delete(w.Name)
+}
+
+func (w *patchWindow) run() (reboot bool) {
+	fmt.Println("DEBUG: run", w.Name)
+
+	w.mx.RLock()
+	defer w.mx.RUnlock()
+	fmt.Println("DEBUG: run got lock", w.Name)
+	defer func() { fmt.Println("DEBUG: run unlock", w.Name) }()
+
+	defer func() {
+		if reboot {
+			if err := saveState(state, w); err != nil {
+				log.Println("ERROR:", err)
+			}
+		} else {
+			if err := saveState(state, nil); err != nil {
+				log.Println("ERROR:", err)
+			}
+		}
+	}()
+
+	// Make sure we are still in the patch window.
+	if !w.in() {
+		fmt.Println(w.Name, "not in patch window")
+		return false
+	}
+
+	fmt.Println("running patch window", w.Name)
+	if err := saveState(state, w); err != nil {
+		log.Println("ERROR:", err)
+	}
+
+	//runUpdates()
+	// Pretend to do work
+	time.Sleep(5 * time.Second)
+
+	// Make sure we are still in the patch window after each step.
+	// TODO: Pass this to runUpdates instead?
+	if !w.in() {
+		fmt.Println(w.Name, "timedout")
+		return false
+	}
+
+	fmt.Println("finished patch window", w.Name)
+	return false
+}
+
+func patchManager(efps []*osconfigpb.LookupConfigsResponse_EffectivePatchPolicy) {
+	// Deregister any existing patchWindows that no longer exist.
+	var ppns []string
+	for _, pp := range efps {
+		ppns = append(ppns, pp.GetFullName())
+	}
+
+	var toDeregister []*patchWindow
+	aw.mx.Lock()
+	for _, w := range aw.windows {
+		if !strIn(w.Name, ppns) {
+			toDeregister = append(toDeregister, w)
+		}
+	}
+
+	fmt.Printf("DEBUG: list of patchWindows to deregister: %q\n", toDeregister)
+	for _, w := range toDeregister {
+		defer w.deregister()
+	}
+
+	// Update or create patchWindows based on provided patch policies.
+	for _, pp := range efps {
+		fmt.Println("DEBUG: setup", pp.GetFullName())
+		if w, ok := aw.windows[pp.GetFullName()]; ok {
+			fmt.Println("DEBUG: patchWindow to update:", w.Name)
+			defer w.update(pp)
+			continue
+		}
+
+		w, err := newPatchWindow(pp)
+		if err != nil {
+			log.Print("ERROR:", err)
+			continue
+		}
+		fmt.Println("DEBUG: patchWindow to create:", w.Name)
+		defer w.register()
+	}
+	aw.mx.Unlock()
```

Comment:
Maybe one way to simplify the locking is to look at it like this: if the agent is running a patch or updating the OsConfig, it doesn't need to call the service or update `aw.windows`. And if it's in the process of updating `aw.windows`, then no patch should start. Not sure if that helps, but it's a mental model that might simplify things.

Reply:
Ack, this is probably way more complicated than it needs to be, a single lock probably can take care of all this as no two things need to happen at once. I'll look at that in the next PR

Reply:
Added todo to look at this once the logic is more complete